File: WebDriverTests_Java/WebDriverTests/src/test/java/core/supportClass/ByNthOfBy.java
package core.supportClass;

import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;

import java.util.Arrays;
import java.util.List;
import java.util.NoSuchElementException;

public class ByNthOfBy extends By
{
    private By by;
    private int index;

    public ByNthOfBy(By by, int index)
    {
        this.by = by;
        this.index = index;
    }

    @Override
    public WebElement findElement(SearchContext context)
    {
        List<WebElement> elements = by.findElements(context);
        if (index < 0 || index >= elements.size())
        {
            throw new NoSuchElementException(String.format("Element with by '%1$s' and index '%2$s' not found (actual count='%3$s')", by, index, elements.size()));
        }
        return elements.get(index);
    }

    @Override
    public List<WebElement> findElements(SearchContext context)
    {
        List<WebElement> elements = by.findElements(context);
        if (index < 0 || index >= elements.size())
        {
            return Arrays.asList(new WebElement[0]);
        }
        return Arrays.asList( new WebElement[] {elements.get(index)});
    }

    @Override
    public String toString()
    {
        return String.format("ByNthOfBy: <%1$s>[%2$s]", by, index);
    }
}

File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/TestSystem/ByExtensions/ByNthOfBy.cs
﻿using System.Collections.ObjectModel;
using OpenQA.Selenium;

namespace Autotests.Utilities.WebTestCore.TestSystem.ByExtensions
{
    public class ByNthOfBy : By
    {
        private readonly By by;
        private readonly int index;

        internal ByNthOfBy(By by, int index)
        {
            this.by = by;
            this.index = index;
        }

        public override IWebElement FindElement(ISearchContext context)
        {
            ReadOnlyCollection<IWebElement> elements = by.FindElements(context);
            if (index < 0 || index >= elements.Count)
                throw new NoSuchElementException(
                    string.Format("Element with by '{0}' and index '{1}' not found (actual count='{2}')", by, index,
                        elements.Count));
            return elements[index];
        }

        public override ReadOnlyCollection<IWebElement> FindElements(ISearchContext context)
        {
            ReadOnlyCollection<IWebElement> elements = by.FindElements(context);
            if (index < 0 || index >= elements.Count)
                return new ReadOnlyCollection<IWebElement>(new IWebElement[0]);
            return new ReadOnlyCollection<IWebElement>(new[] {elements[index]});
        }

        public override string ToString()
        {
            return string.Format("ByNthOfBy: <{0}>[{1}]", by, index);
        }
    }
}